In [1]:
from IPython.display import display

from sympy import *
import numpy as np

# x, y, z = symbols('x y z')
init_printing(use_unicode=True)

# Homework

## Exercise set 4.1

### 2a

Use the forward-difference forumlas and backward-difference formulas to determine each missing entry in the following tables.

|$x$|$f(x)$|$f'(x)$|
|---|---|---|
|-0.3|1.9507||
|-0.2|2.0421||
|-0.1|2.0601||

**solution**

|$x$|$f(x)$|$f'(x)$|
|---|---|---|
|-0.3|1.9507|0.9140|
|-0.2|2.0421|0.1800|
|-0.1|2.0601|0.1800|

In [2]:
h = 0.1
# forward
print((2.0421 - 1.9507) / h)
print((2.0601 - 2.0421) / h)

h = -0.1

# backward
print((1.9507 - 2.0421) / h)
print((2.0421 - 2.0601) / h)

0.9139999999999993
0.17999999999999794
0.9139999999999993
0.17999999999999794


### 4a

The data in Exercise 2 were taken from the following functions.
Compute the actual errors in Exercise 2, and find error bounds using the error formulas.

$f(x)=2\cos 2x-x$

**solution**

$f'(x)=-4\sin 2x - 1$

$f''(x)=-8\cos 2x$

$$\frac{|hf''(\xi)|}{2}=\frac{|-8h\cos 2x|}{2}=|4hcos2x|$$

$-0.3 < -\xi < -0.2$

$f'(-0.3)=1.2586$, Absolute error: 0.3446 , Error bound < 0.3684

$-0.2 < -\xi < -0.1$

$f'(-0.2)=0.5577$, Absolute error: 0.3777 , Error bound < 0.3920

$-0.2 < -\xi < -0.1$

$f'(-0.1)=-0.2053$, Absolute error: 0.3853 , Error bound < 0.3920

In [3]:
f = lambda v: -4 * sin(2 * v) - 1
v = -0.3
vp = round(f(v), 4)
rel = max(abs(4 * h * cos(2 * v)), abs(4 * h * cos(2 * (v + h))))
print(f'f\'({v}) =', vp, ', Absolute error:', round(abs(vp - 0.914), 4), ', Error bound <', round(rel, 4))
v = -0.2
vp = round(f(v), 4)
rel = max(abs(4 * h * cos(2 * v)), abs(4 * h * cos(2 * (v + h))))
print(f'f\'({v}) =', vp, ', Absolute error:', round(abs(vp-0.18), 4), ', Error bound <', round(rel, 4))
v = -0.1
vp = round(f(v), 4)
rel = max(abs(4 * h * cos(2 * v)), abs(4 * h * cos(2 * (v - h))))
print(f'f\'({v}) =', vp, ', Absolute error:', round(abs(vp-0.18), 4), ', Error bound <', round(rel, 4))

f'(-0.3) = 1.2586 , Absolute error: 0.3446 , Error bound < 0.3301
f'(-0.2) = 0.5577 , Absolute error: 0.3777 , Error bound < 0.3684
f'(-0.1) = -0.2053 , Absolute error: 0.3853 , Error bound < 0.4000


### 6a

Use the most accurate three-point formula to determine each missing entry in the following tables.

|$x$|$f(x)$|$f'(x)$|
|---|---|---|
|-0.3|-0.27652| |
|-0.2|-0.25074| |
|-0.1|-0.16134| |
|0|0| |

**solution**

|$x$|$f(x)$|$f'(x)$|
|---|---|---|
|-0.3|-0.27652| -0.0603|
|-0.2|-0.25074| 0.5759|
|-0.1|-0.16134| 1.2537|
|0|0| 1.9731|

In [4]:
h = 0.1
f = [-0.27652, -0.25074, -0.16134, 0]

# endpoint
endpoint = lambda i: 1/(2 * h) * (-3 * f[i] + 4 * f[i + 1] - f[i + 2])

# midpoint
midpoint = lambda i: 1/(2 * h) * (f[i + 1] - f[i - 1])

# -0.3
print(round(endpoint(0), 5))

# -0.2
print(round(midpoint(1), 5))

# -0.1
print(round(midpoint(2), 5))

# 0
h = -0.1
endpoint = lambda i: 1/(2 * h) * (-3 * f[i] + 4 * f[i - 1] - f[i - 2])
print(round(endpoint(3), 5))

-0.0603
0.5759
1.2537
1.9731


### 8a

The data in Exercise 6 were taken from the following functions.
Compute the actual errors in Exercise 6, and find error bounds using the error formulas.

$f(x)=e^{2x}- \cos 2x$

**solution**

$f'(x)=2(e^{2x} + \sin 2x)$

$f''(x)=4(e^{2x} + \cos 2x)$

$$\frac{|hf''(\xi)|}{2}=\frac{|4h(e^{2x} + \cos 2x)|}{2}=|2h(e^{2x} + \cos 2x)|$$

$f'(-0.3) = -0.0317$ , Absolute error: 0.0286 , Error bound < 0.6366

$f'(-0.2) = 0.5618$ , Absolute error: 0.0141 , Error bound < 0.7195

$f'(-0.1) = 1.2401$ , Absolute error: 0.0136 , Error bound < 0.8000

$f'(0) = 2$ , Absolute error: 0.0269 , Error bound < 0.8000


In [5]:
h = 0.1
fp = lambda X: 2 * (exp(2 * X) + sin(2 * X))
fpp = lambda X: 4 * (exp(2 * X) + cos(2 * X))
for X, est in zip([-0.3, -0.2, -0.1, 0], [-0.0603, 0.5759, 1.2537, 1.9731]):
    rel = max(fpp(X), fpp(X + h)) if X != 0 else max(fpp(X), fpp(X - h))
    print(f'$f\'({X}) = {round(fp(X), 4)}$', ', Absolute error:', round(abs(fp(X) - est), 4), ', Error bound <', round(rel * h, 4))

$f'(-0.3) = -0.0317$ , Absolute error: 0.0286 , Error bound < 0.6366
$f'(-0.2) = 0.5618$ , Absolute error: 0.0141 , Error bound < 0.7195
$f'(-0.1) = 1.2401$ , Absolute error: 0.0136 , Error bound < 0.8000
$f'(0) = 2$ , Absolute error: 0.0269 , Error bound < 0.8000


## Exercise set 4.2

### 1c

Apply the extrapolation process described in Example 1 to determine $N_3(h)$, an approximation to $f'(x_0)$, for the following functions and stepsizes.

$f(x)=2^x \sin x, x_0=1.05, h=0.4$

**solution**

$f'(x)=2^x (\cos x + \ln 2 \sin x)$

$f'(1.05) = 2.2751458417295467$

$N_3(1.05) = 2.27515$

In [58]:
f = lambda X: np.power(2, X) * np.sin(X)
fp = lambda X: np.power(2, X) * (np.cos(X) + np.log(2) * np.sin(X))
print(fp(1.05))

# https://www.marksmath.org/classes/Spring2016NumericalAnalysis/demos/RichardsonExtrapolation.html
# center diff
def phi(X,h): return (f(X+h)-f(X-h))/(2*h)

d = [phi(1.05, h) for h in [0.4 * 2**(-n) for n in range(3)]]
print(d)

N = len(d)
D = np.zeros((N,N))
D[:,0] = d
for m in range(1,N):
    for n in range(m,N):
        D[n,m] = (4**m*D[n,m-1]-D[n-1,m-1])/(4**m-1)
D

2.2751458417295467
[2.2031656973915594, 2.257237243413355, 2.2706741673063457]


array([[2.2031657 , 0.        , 0.        ],
       [2.25723724, 2.27526109, 0.        ],
       [2.27067417, 2.27515314, 2.27514595]])

## Exercise set 4.3

### 2a

Approximate the following integrals using the Trapezoidal rule.

$$\int_{-0.25}^{0.25} (\cos x)^2 dx$$

**solution**

$$\int_{-0.25}^{0.25} f(x) dx \approx \frac{(0.25 + 0.25)}{2}(f(-0.25) + f(0.25)) \approx 0.470$$

In [81]:
f = lambda X: np.power(np.cos(X), 2)
(0.5/2) * (f(-0.25) + f(0.25))

0.46939564047259336

### 4a

Find a bound for the error in Exercise 2 using the error formula, and compare this to the actual error.

**solution**

Actual error: $0.489713 - 0.46940 < 0.021$

Error bound:

$f''(x) = 2\sin^2 x - 2\cos^2 x$

$$\frac{h^3}{12}f''(\xi)=\frac{h^3}{6}\sin^2 \xi \cos^2 \xi \approx 0.0012$$

I'm missing something in the above calculation, or my original estimate is wrong.

In [77]:
# abs err
print(0.489713 - 0.46939564)

(np.power(0.5, 3) / 6) * (pow(np.sin(0.25), 2) * pow(np.cos(0.25), 2))

0.020317360000000007


0.0011971294118017202

### 6a

Repeat Exercise 2 using Simpson's rule.

**solution**

$$\int_{-0.25}^{0.25} f(x) dx \approx \frac{(0.25 + 0.25)}{3}(f(-0.25) + 4f(0)+ f(0.25)) \approx 0.490$$

In [87]:
f = lambda X: np.power(np.cos(X), 2)
h = 0.25
(h / 3) * (f(-0.25) + 4 * f(0) + f(0.25))

0.4897985468241977